### PR TITLE
Serialization

### PR DESCRIPTION
**Pull request by whooo**
Adding functionality to be able to serialize different TPM structures

### CODE REVIEW

**Comment by whooo** [comment]
While this doesn't exactly cover https://github.com/tpm2-software/tpm2-pytss/issues/235 it might be useful as a base.

File: tpm2_pytss/constants.py [hunk 1]
```diff
@@ -1267,6 +1310,16 @@ class TPMA_CC(TPM_FRIENDLY_INT):
     RES_MASK = lib.TPMA_CC_RES_MASK
     RES_SHIFT = lib.TPMA_CC_RES_SHIFT
 
+    @property
+    def commandindex(self) -> int:
+        """int: The command index"""
+        return self & self.COMMANDINDEX_MASK
+
+    @property
+    def chandles(self) -> int:
```
Comment (williamcroberts) [inline tpm2_pytss/constants.py:1319]: do we want:

```python
def is_flushed()
def is_vendor()
def is_nv()
def is_extensive()
def has_rhandle()
... etc ...
```

Comment (whooo) [reply tpm2_pytss/constants.py:1322]: Possibly, could be useful with the different bits of TPMA_NV and TPMA_OBJECT for example

File: test/test_encoding.py [hunk 2]
```diff
@@ -0,0 +1,743 @@
...
+    base_encdec,
+    toolslike_encdec,
+    json_encdec,
+)
+
+
+class SerializationTest(unittest.TestCase):
+    def test_base_simple_tpm2b(self):
+        enc = base_encdec()
+        dig = TPM2B_DIGEST(b"falafel")
```
Comment (williamcroberts) [inline test/test_encoding.py:18]: im hungry now

File: test/test_encoding.py [hunk 3]
```diff
@@ -0,0 +1,743 @@
+#!/usr/bin/python3 -u
+"""
+SPDX-License-Identifier: BSD-2
+"""
+import unittest
+
+from tpm2_pytss import *
+from tpm2_pytss.encoding import (
+    base_encdec,
```
Comment (williamcroberts) [inline test/test_encoding.py:9]: So what's base_encdec all hex strings? If so perhaps call this hex_encdec

File: test/test_encoding.py [hunk 4]
```diff
@@ -0,0 +1,743 @@
...
+    toolslike_encdec,
```
Comment (williamcroberts) [inline test/test_encoding.py:10]: Is it supposed to be tools like or actually match the tools? If it should match the tools, lets not be ambiguous about that and call it tpm2tools_encdec.

Plus pep8 classes should be camelCased.

File: test/test_encoding.py [hunk 5]
```diff
@@ -0,0 +1,743 @@
...
+    json_encdec,
```
Comment (williamcroberts) [inline test/test_encoding.py:11]: I'm assuming this matches the json spec?
  - https://trustedcomputinggroup.org/resource/tcg-tss-json/

File: test/test_encoding.py (hunk 3 above)
Comment (whooo) [reply test/test_encoding.py:9]: it's the base class for other encoders/decoders, so it's the minimal viable implementation so subclasses can just implement the needed changes

File: test/test_encoding.py (hunk 5 above)
Comment (whooo) [reply test/test_encoding.py:11]: That's the idea

File: test/test_encoding.py (hunk 4 above)
Comment (williamcroberts) [reply test/test_encoding.py:10]: Whats the difference, ordering or missing fields?

File: test/test_encoding.py (hunk 3 above)
Comment (williamcroberts) [reply test/test_encoding.py:9]: Is it supposed to be public? If it's public the name should be more descriptive of what it does IMO.

File: test/test_encoding.py (hunk 4 above)
Comment (whooo) [reply test/test_encoding.py:10]: different field names and tpm2-tools flatten some structures, for example tpm2_readpublic puts the RSA parameters at the same level as type and nameAlg, so toolslike_encdec should be seen more as TPMA_FRIENDLY_INTLIST.parse then a compatibility layer for the tools

Comment (williamcroberts) [reply test/test_encoding.py:10]: > different field names and tpm2-tools flatten some structures, for example tpm2_readpublic puts the RSA parameters at the same level as type and nameAlg, so toolslike_encdec should be seen more as TPMA_FRIENDLY_INTLIST.parse then a compatibility layer for the tools

Folks over in tpm.dev are clamoring for a way to get tools YAML in and out to match. Is their anyway we can add in the special handling for tools?

File: test/test_encoding.py (hunk 3 above)
Comment (whooo) [reply test/test_encoding.py:9]: I would prefer it to be public, if users want to implement their own encoders/decoders
I could always rename it to minimal_encdec and improve the description in the docstring

File: test/test_encoding.py (hunk 4 above)
Comment (whooo) [reply test/test_encoding.py:10]: It's possible, I wouldn't do that in this class tho
It's just a lot of work going through all the source code for the tools and implementing + verifying it

Comment (williamcroberts) [reply test/test_encoding.py:10]: > It's possible, I wouldn't do that in this class tho It's just a lot of work going through all the source code for the tools and implementing + verifying it

Let's take base and json now, and then start a longer running branch for tools.

Comment (williamcroberts) [reply test/test_encoding.py:10]: FYI the tools YAML support is indicated in ticket: https://github.com/tpm2-software/tpm2-pytss/issues/235

Comment (whooo) [reply test/test_encoding.py:10]: > 
> Let's take base and json now, and then start a longer running branch for tools.

I would like to keep toolslike as well, as I have at least two use cases for that
I'll see if I can create a new branch of this and start the tools-part there in a couple of days

Comment (williamcroberts) [reply test/test_encoding.py:10]: > > Let's take base and json now, and then start a longer running branch for tools.
> 
> I would like to keep toolslike as well, as I have at least two use cases for that I'll see if I can create a new branch of this and start the tools-part there in a couple of days

The problem with tools-like, is that if we merge it and don't make it tools-firm by release, we have effectively made a new format. We already have enough formats to deal with. If you want to hide it behind a feature flag where to use it you have to buy into the fact it's experimental, that's fine too. You could just make it _toolslike_encdec indicating that it's private still and you probably shouldn't use it, unless you know what your doing, like you do.

Comment (whooo) [reply test/test_encoding.py:10]: I see your point, I'll remove it

**Review by williamcroberts — APPROVED** [review]

**Pull request merged**